utils/initial_data_preparation.py

```diff
@@ -160,71 +190,103 @@ def is_complete_url(url):
-    incomplete_urls = df[~df["repourl"].apply(is_complete_url)]
+    non_duplicate_rows = ~df["duplicate_flag"]  # Identify non-duplicate rows
+    domain_extraction_successful = ~df["unsupported_url_scheme"]
 
-    # Log the incomplete URLs
-    if not incomplete_urls.empty:
-        logger.warning(
-            f"{len(incomplete_urls)} incomplete GitHub URLs found and will not "
-            f"be analysed:"
-        )
-        for url in incomplete_urls["repourl"]:
-            logger.info(f"Excluding the repourl: {url}")
+    # Check for successful domain extraction
+    # Apply the incomplete URL flag condition on non-duplicate and unsuccessful
+    # domain extraction rows
+    df["incomplete_url_flag"] = ~df.loc[
+        non_duplicate_rows & domain_extraction_successful, "repourl"
+    ].apply(is_complete_url)
+    df["incomplete_url_flag"] = df["incomplete_url_flag"].astype(bool)
+    incomplete_count = df["incomplete_url_flag"].sum()
+    logger.info(f"Found {incomplete_count} incomplete " f"URLs.")
 
-    # Filter out incomplete URLs using the helper function
-    filtered_df = df[df["repourl"].apply(is_complete_url)]
-    return filtered_df
+    return df
 
 
-def get_base_repo_url(url):
+def get_base_repo_url(df):
     """
-    Extracts the base repository URL from various hosting platforms.
-    Handles URLs ending with '.git', supports nested groups in GitLab,
-    and adapts to different platform URL structures.
+    Extracts the base repository URL from various hosting platforms and logs
+    the process.
+
+    Extracts the base repository URL from various hosting platforms and adds a
+    'base_repo_url_flag' column to indicate success or failure of extraction.
+    Returns None immediately if the dataframe is None or empty. Determines
+    the base path based on the hosting platform and URL structure.
+    Filters DataFrame based on specified conditions (non-duplicate rows,
+    supported URL schemes, and complete URLs). Logs the process and counts
+    the rows with URL extraction issues.
 
     Args:
-      url (str): The full URL to a Git repository.
+       df (pd.DataFrame): DataFrame containing URLs.
 
     Returns:
-      str: The base repository URL if valid, otherwise returns None.
+      pd.DataFrame: The updated DataFrame with 'base_repo_url' and
+      'base_repo_url_flag' columns.
     """
-    # Return None immediately if the URL is None or empty
-    if not url:
+
+    # Return None immediately if the dataframe is None or empty
+    if df.empty:
         return None
 
-    parsed_url = urlparse(url)
-    path = parsed_url.path.strip("/")
+    def extract_url(url):
+        if not url:
+            return None, True  # Return None for URL, True for unsuccessful flag
 
-    parts = path.split("/")
-    print(f"parts: {parts}")
+        parsed_url = urlparse(url)
+        # Strip leading '/' to avoid the first empty string in the list
+        path = parsed_url.path.strip("/")
```

Comment:
Given our discussion about possible bugs in removing '/' characters incorrectly at times, could this be tightened up to remove only the '/' if it is the leading character(s). 

FWIW using the interactive python editor https://www.w3schools.com/python/ref_string_lstrip.asp we can quickly see the effects of the 3 variations of the strip function.

```
txt = "//domain.com/owner/repo/"
print(txt.lstrip('/'))
print(txt.strip('/'))
print(txt.rstrip('/'))
```

Reply:
Given there are `repourls` like `https://github.com/inters/vita/` which have `/` at the beginning and end, `strip` would remove both of them.Please see below:
`parsed_url = urlparse("https://github.com/inters/vita/")`
`path = parsed_url.path.strip("/")`
`parts = path.split("/")` --> ['inters', 'vita']

But if `lstrip` is used: `parts` would become `['inters', 'vita', '']` which then makes the rest of the actions wrong (base_path = "/".join(parts[:2]) )